### PR TITLE
Leverage metadata to populate gems urls

### PR DESCRIPTION
**Pull request by kbrock**
This parses common urls from the `gemspec` metadata. It uses the same keys as the api (found in `gemspec.rb`).

Addresses questions I've seen in mailing list and gh issues.

Addresses issues in #718

Thanks

/cc @qrush @fwolfst

**UPDATE:** reduced heading comment since the original concern was determining the keys to use. (this was resolved by using the ones in `gemspec.rb`

### CODE REVIEW

**Comment by kbrock** [comment]
@sferik If this is good, should pull requests like #728 follow a similar route in adding to the metadata?

File: test/helpers/gem_helpers.rb [hunk 1]
```diff
@@ -17,7 +17,7 @@ def build_gem(name, version, summary = "Gemcutter", platform = "ruby", &block)
     build_gemspec(new_gemspec(name, version, summary, platform, &block))
   end
 
-  def new_gemspec(name, version, summary, platform, &block)
+  def new_gemspec(name, version, summary, platform, metadata = nil, &block)
```
Comment (bf4) [inline test/helpers/gem_helpers.rb]: wait, why do you default `metadata = nil`? I should be `{}`, no?

File: app/models/linkset.rb [hunk 2]
```diff
@@ -12,6 +12,14 @@ def empty?
...
-    self.update_attributes!(:home => spec.homepage)
+    meta = spec.metadata || {}
+    self.update_attributes!({
+      :home => spec.homepage,
+      :wiki => meta["wiki_uri"],
+      :docs => meta["documentation_uri"],
+      :mail => meta["mailing_list_uri"],
+      :code => meta["source_code_uri"],
+      :bugs => meta["bug_tracker_uri"]
+    }.delete_if { |_n, v| v.nil? })
```
Comment (bf4) [inline app/models/linkset.rb]: Is there a corresponding PR in rubygems to add these keys as recommended? If think those are fine metadata names.  Is Discourse a mailing list?  What about IRC?  Twitter? Contact Email? Something Freeform?

Comment (mpapis) [inline app/models/linkset.rb]: the `metadata` is freeform, but the problem is it is limited, you can only define strings and with limited size, anothe thing is those urls are already handled, so it's basically new way of specifying this urls, if you want more data then it would be part of new request

File: test/helpers/gem_helpers.rb (hunk 1 above)
Comment (simi) [inline test/helpers/gem_helpers.rb]: :+1: for `{}`

Comment (mpapis) [inline test/helpers/gem_helpers.rb]: check the next added line:
```ruby
s.metadata = metadata if metadata
```
if it would be `{}` then it could reset any defaults ... although there aren't any

Comment (simi) [inline test/helpers/gem_helpers.rb]: My idea behind this was to remove responsibility for this in `app/models/linkset.rb` and remove `meta = spec.metadata || {}` line. I think this is best place to take responsibility.

File: app/models/linkset.rb (hunk 2 above)
Comment (kbrock) [inline app/models/linkset.rb]: @bf4 This started from #718 - there are a lot of possibilities here and a lot of different projects that need to be updated.

Since we return these keys in the api, this is allowing developers to populate it from the gemspec metadata. You can see api defined in [rubygems](https://github.com/rubygems/rubygems.org/blob/63d43b660aa8430320c0db430b15e2def306e3ac/app/models/rubygem.rb#L162-L167).

File: test/helpers/gem_helpers.rb (hunk 1 above)
Comment (kbrock) [inline test/helpers/gem_helpers.rb]: @bf4 @simi As stated [before][], A parsed gemspec can return a nil `metadata` hash. This is in `Gem::Package.new("filename").spec` and not part of the rubygems code base.

Changing `linkset.rb` and `new_gemspec` will pass for tests using the `new_gemspec` factory, but blows up when parsing real gems.

[before]: https://github.com/rubygems/rubygems.org/pull/724#discussion_r19806413

File: test/unit/linkset_test.rb [hunk 3]
```diff
@@ -35,4 +35,45 @@ class LinksetTest < ActiveSupport::TestCase
       assert_equal @spec.homepage, @linkset.home
     end
   end
+
+  context "with a previously populated linkset" do
+    setup do
+      @linkset = create(:linkset,
+             :wiki => "http://example.com/old_wiki",
```
Comment (simi) [inline test/unit/linkset_test.rb]: I'm not getting this indentation.

Comment (kbrock) [inline test/unit/linkset_test.rb]: sweet - thanks.

aka I'm not getting this indentation either

**Comment by davefp** [comment]
@kbrock As a result of #610 a large chunk of the repo's history was re-written. Please rebase your PR so that it will merge cleanly. Thanks!

(There are further details on the blog: http://blog.rubygems.org/2015/02/01/rewriting-history.html)

File: app/models/linkset.rb [hunk 4]
```diff
@@ -13,6 +13,14 @@ def empty?
   end
 
   def update_attributes_from_gem_specification!(spec)
-    self.update_attributes!(:home => spec.homepage)
+    meta = spec.metadata || {}
```
Comment (kbrock) [inline app/models/linkset.rb]: Final answer: If there is no `metadata` in the `Gemfile`, `spec.metadata == nil`. This safeguards it.

**Comment by kbrock** [comment]
@davefp / @arthurnn Rebased. Anything I can do to push this forward?

Thanks

**Comment by kbrock** [comment]
Rebased again to allow a clean merge. Let me know if there is anything else to add to help get this merged

**Comment by kbrock** [comment]
Suggestions how to push this forward?
Thanks

**Comment by qrush** [comment]
@kbrock I'd really prefer to get #923 deployed and deal with the support queue backlog before adding more features.

I'll get you invited to the Bundler/rubygems.org Slack so you can keep up to date (and that goes for anyone else interested as well).

**Comment by vassilevsky** [comment]
I'm interested. Could you invite me? My email is vassilevsky@gmail.com .

**Comment by mpapis** [comment]
Could you invite me too, my email is mpapis+rubygems@gmail.com

**Comment by JuanitoFatas** [comment]
Could you invite me too? My email is `katehuang0320@gmail.com`, thanks! :balloon:

**Comment by kbrock** [comment]
Anything special I can do to help this forward.

I'm thinking that no one will notice until people start adding meta-data.
(I just pushed a gem through with this meta data and a changelog reference in it)

**Comment by arthurnn** [comment]
We persist the metadata now on the gem version . https://github.com/rubygems/rubygems.org/pull/858 . Not sure if that changes the behaviour of this or not.

**Comment by kbrock** [comment]
@arthurnn thanks. That does change things. I'll try and get back in there over the weekend

**Comment by kbrock** [comment]
@qrush I can rewrite this to read the urls from the metadata and fallback on the links table.
Once more gems are using the metadata, we can remove the links table.

It won't be as simple as +9/-1, but it should be close.

We could also change the ui to display any arbitrary link instead of only the pre-prescribed ones. I just worry about lack of standards and names that don't mean anything.

**Comment by arthurnn** [comment]
> I can rewrite this to read the urls from the metadata and fallback on the links table.
Once more gems are using the metadata, we can remove the links table.

I think this is a good start.

**Comment by kbrock** [comment]
This did not address education in the edit form. We can do anything from just displaying a friendly note, to hiding fields that are set in versions, to removing the form all together.

Thanks

**Comment by kbrock** [comment]
@arthurnn / @mpapis I updated to take into account the stored metadata. Does this look like it still meets our needs?

File: app/views/rubygems/show.html.erb [hunk 5]
```diff
@@ -147,16 +147,13 @@
         <%= link_to t('edit'), edit_rubygem_path(@rubygem), :class => "gem__link t-list__item", :id => "edit" if @rubygem.owned_by?(current_user) %>
 
         <% if @latest_version.indexed %>
-          <% if @rubygem.linkset.present? %>
-            <%- %i{home code docs wiki mail bugs}.each do |link| %>
-              <%= link_to_page t("rubygems.show.links.#{link}"), @rubygem.linkset.public_send(link) %>
-            <%- end %>
-          <% end %>
+          <%- Linkset::LINKS.each do |link, attribute| %>
```
Comment (olivierlacan) [inline app/views/rubygems/show.html.erb]: Big fan of this, literally just suggested this in the original commit today. :+1:

File: app/models/version.rb [hunk 6]
```diff
@@ -309,6 +309,11 @@ def recalculate_metadata!
     nil
   end
 
+  def link_uri(name, linkset = nil)
+    metadata[name.to_s].presence || linkset.try(name).presence ||
+      (name.to_s == "documentation_uri" ? documentation_path : nil)
```
Comment (arthurnn) [inline app/models/version.rb]: instead of this condition, can we add the `documentation_uri` method on the linkset?

File: test/functional/rubygems_controller_test.rb [hunk 7]
```diff
@@ -28,6 +28,7 @@ class RubygemsControllerTest < ActionController::TestCase
         @owners = [@user, create(:user)]
         @rubygem = create(:rubygem, owners: @owners, number: "1.0.0")
         @rubygem.linkset = nil
+        @latest_version = @rubygem.versions.latest
```
Comment (arthurnn) [inline test/functional/rubygems_controller_test.rb]: why was this necessary?

Comment (kbrock) [inline test/functional/rubygems_controller_test.rb]: +1, removed.

File: app/models/version.rb (hunk 6 above)
Comment (kbrock) [inline app/models/version.rb]: It would be nice to remove this and read better.

`linkset#documentation_uri` has access to `rubygem#name`, but does not have access to `version#number`. Ideas?

```ruby
  def documentation_path
    "http://www.rubydoc.info/gems/#{rubygem.name}/#{number}"
  end
```

Comment (kbrock) [inline app/models/version.rb]: I have fixed this in the alt solution. thanks for raising this concern

**Comment by kbrock** [comment]
@arthurnn did you get a chance to look at #1234? do you find that code simpler / better?
thanks

**Comment by kbrock** [comment]
closing, preferring #1234
not worth the constant rebasing
